### PR TITLE
Add selective activation checkpointing (SAC) to SFT

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Adds selective activation checkpointing (SAC) to SFT. With gradient checkpointing on, the attention output is saved during forward instead of being recomputed in backward, recovering most of the checkpointing slowdown at long context for one extra hidden-state-sized tensor per layer. Same eager SAC approach torchtitan uses under FSDP2, no `torch.compile` needed.

Turn it on with `selective_activation_checkpointing=True` on `SFTConfig` (needs `gradient_checkpointing=True`). PEFT + DeepSpeed ZeRO-3 is rejected since that combo needs reentrant checkpointing, which SAC can't use.

Split out of #6280, addressing review: `enable_selective_activation_checkpointing` now lives in its own module instead of `activation_offloading.py`, the op-matching policy is unit-tested directly instead of by counting SDPA calls in an actual forward/backward, and the test class requires an accelerator.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

cc @qgallouedec @winglian

---

> [!NOTE]
> **Medium Risk**
> Changes gradient checkpointing behavior on the training path and can affect memory/performance; incompatible distributed configs are rejected explicitly, but wrong flag combinations could surprise users until they hit validation errors.
> 
> **Overview**
> Adds **selective activation checkpointing (SAC)** for [`SFTTrainer`], opt-in via `SFTConfig(selective_activation_checkpointing=True)` with `gradient_checkpointing=True`. A new `enable_selective_activation_checkpointing` helper wraps the model’s `gradient_checkpointing_enable` so PyTorch’s selective checkpoint policy **saves attention outputs** (SDPA `aten` ops and `flash_attn` kernels) while **recomputing other ops**, using non-reentrant checkpointing only.
> 
> `SFTTrainer` validates SAC prerequisites, applies the wrapper before `Trainer` init, and **errors** on incompatible setups (SAC without gradient checkpointing, `use_reentrant=True`, or **PEFT + DeepSpeed ZeRO-3**). Tests cover the op policy and gradient parity vs full checkpointing; the paper index documents the feature.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9cf7043f8a2f9fa6a5edf08bbf81629be2826968. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — COMMENTED** [review]
All three points from #6280 addressed 👌

Validated myself
Counted `scaled_dot_product` dispatches on `tiny-Qwen2ForCausalLM-2.5` (2 layers), one fwd+bwd: 6 with full checkpointing, 4 with SAC. Three per layer vs two, the extra being the recompute. So `sdpa_full - sdpa_sac == num_hidden_layers` is deterministic and fails the moment the context function stops applying. Worth adding, it's what separates the policy being correct from the policy being used.

One suggestion: I think at some point we will probably upstream this to transformers. And if we do, using `gradient_checkpointing_kwargs` is probably more aligned with the conventions. Instead of adding a new config, maybe we can simply pop `selective_activation_checkpointing` from `gradient_checkpointing_kwargs`?

File: tests/test_selective_activation_checkpointing.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from trl.models.selective_activation_checkpointing import (
+    _aten_attention_ops,
+    _build_policy_fn,
+    enable_selective_activation_checkpointing,
+)
+
+from .testing_utils import TrlTestCase, require_torch_accelerator
+
+
+class TestSelectiveActivationCheckpointingPolicy(TrlTestCase):
```
Comment (qgallouedec) [inline tests/test_selective_activation_checkpointing.py:29]: I think nothing tests the policy is actually wired in. Drop `context_fn` and all four tests still pass, since `test_matches_full_checkpointing` compares SAC against full checkpointing and a silent degradation makes the gradients match trivially.
